### PR TITLE
remove patches and add exit status

### PR DESCRIPTION
**Pull request by ocefpaf**
I got one user reporting that the command

```
ncap2 -O -v -s 'erf_one=float(gsl_sf_erf(1.0f));print(erf_one,"%g")' in.nc foo.nc
```

was segfaulting.  I added the `` || exit 1`` b/c `AppVeyor` may have failed silently on us in #48

### CODE REVIEW

**Comment by czender** [comment]
The AppVeyor log that was sent to me shows that the format argument to print() is "g" not "%g" (below). However the test command in recipes/run_test.bat appears to have "%g" which is correct. Maybe this is just a Windows printing artifact, or maybe it is indicative of a problem. UNIX works fine with either argument.

`(root) C:\Miniconda36-x64\conda-bld\nco_1513461168463\test_tmp\data>ncap2 -O -v -s 'erf_one=float(gsl_sf_erf(1.0f));print(erf_one,"g")' in.nc foo.nc   || exit 1
`

**Comment by czender** [comment]
Sorry, I got confused. This PR is for nco-feedstock not for nco proper, so I can't merge anything :)

**Comment by ocefpaf** [comment]
> Sorry, I got confused. This PR is for nco-feedstock not for nco proper, so I can't merge anything :)

I am trying to find out what is wrong with `conda-forge`'s `gsl` before merging this. The PR is only to show the error. BTW, users reported that either `%g` or `g` triggers the segfault on Windows.

**Comment by ocefpaf** [comment]
It seems that we get the same failure. See https://github.com/conda-forge/nco-feedstock/pull/55/commits/20154a7fa0e0c6f057a7f8ee76320ebc07ee426c

**Comment by czender** [comment]
What's strange is that, IIRC @ocefpaf obtained the correct numerical results (0.842701) from the test command
`ncap2 -O -v -s 'erf_one=float(gsl_sf_erf(1.0f));print(erf_one,"%g")' in.nc foo.nc`
 but the results would not print on Windows. So the numerical core of GSL is working (yes?) though segfaulting. It would be helpful to have a stack trace of the segfault...but I don't have Windows access. Maybe @ajcse1 or @pedro-vicente can run this command on Windows and see where in the source code it dies...

**Comment by pedro-vicente** [comment]
this is the output I have on the command line that causes the segfault

`ncap2 -O -v -s 'erf_one=float(gsl_sf_erf(1.0f));print(erf_one,"g")' in.nc foo.nc`

is it possible for a user to modify settings somewhere so that debug and static linking is used on the build? 
my recommendation is to make the default build debug and static 
from my experience, using DLLs is a recipe for disaster... 

here's a compiler warning

`c:\users\pedro\anaconda3\conda-bld\nco_1513638953890\work\src\nco++\fmc_all_cls.hh(49): warning C4273: 'erfcf': inconsistent dll linkage`

**Comment by czender** [comment]
Thanks @pedro-vicente. I have a vague idea of what could be causing the problem. My hypothesis is that this command will work without segfaulting:
`ncap2 -O -v -s 'erf_one=float(gsl_sf_erf(1.0));print(erf_one,"%g")' in.nc foo.nc`
This changes the argument of gsl_sf_erf() from float to double. If this works then I have a good idea what the problem is. Could someone with Windows access please test that command?

**Comment by isuruf** [comment]
Doesn't work. https://ci.appveyor.com/project/isuruf/nco-feedstock/build/1.0.9

**Comment by ocefpaf** [comment]
> Doesn't work. https://ci.appveyor.com/project/isuruf/nco-feedstock/build/1.0.9

Did the same in my AppVeyor instance. Same failure...

**Comment by czender** [comment]
This is a baffling failure. Until porting to WIN64 with conda we had never ported NCO to a platform where sizeof(long) = 4 != 8 = sizeof(size_t). We thought we addressed all places in the code where this mattered, yet we could have missed some in ncap2. Seems like a GDB stacktrace would lead straight to the problem. We need someone to run a debugger on this command in a WIN64 environment.

**Comment by pedro-vicente** [comment]
@czender 
I tried 
`ncap2 -O -v -s 'erf_one=float(gsl_sf_erf(1.0));print(erf_one,"%g")' in.nc foo.nc`

and it segfaults

**Comment by pedro-vicente** [comment]
@czender 
 

> Seems like a GDB stacktrace would lead straight to the problem. 

yes, exactly. But for that I need a debug build

@ocefpaf 
I tried to modify bld.bat so that the NCO Cmake builds in "debug", but this causes link errors because  the dependencies were build with different settings, release maybe?

Is there a way to force the download of the debug libraries, if they exist? If they don't exist, is there a way to build all in debug mode?

**Comment by ocefpaf** [comment]
> Is there a way to force the download of the debug libraries, if they exist? If they don't exist, is there a way to build all in debug mode?

In `conda-forge` we do not ship libraries in debug mode. I guess that they are slower and bigger than the non-debug version (is that correct @isuruf?).

I am planning on re-building all the stack in debug mode this weekend (not sure how hard that will be).

If you want to give it a shot all the recipes are in the conda-forge organization. You would need to build the dependencies locally, in order, with debug mode on.

**Comment by czender** [comment]
@pedro-vicente Before there was conda, there was your private CMake build on Windows32. Did you ever build privately on Win64? If so does the segfault occur there, or is it limited to the Conda build? Perhaps easier to debug outside of Conda on your private machine.

**Comment by pedro-vicente** [comment]
@ocefpaf 

> You would need to build the dependencies locally, in order, with debug mode on.

ok, how can I do this? just modify the Cmake script to use "debug" ?

**Comment by pedro-vicente** [comment]
@czender 

`Did you ever build privately on Win64?`

I never built all the Win64 dependencies, because they are too many to maintain manually. Conda-forge is perfect for this, because it's an automated process

**Comment by czender** [comment]
That makes perfect sense. One question I'm unclear on: does the GSL test command work on your private Win32 build? Or does it seem to be a Win64-specific issue?

**Comment by ocefpaf** [comment]
> ok, how can I do this? just modify the Cmake script to use "debug" ?

That is what I would do but to be honest I was going to ask for help of Windows experts like yourself :smile: 

> I never built all the Win64 dependencies, because they are too many to maintain manually.

Any reason why you chose Win-32 instead of Win-64? We can do both in `conda-forge` but we have plans to abandon Win-32 altogether in the near future,

> That makes perfect sense. One question I'm unclear on: does the GSL test command work on your private Win32 build? Or does it seem to be a Win64-specific issue?

That would be nice to know so we can be sure it is a conda+Win-64 issue.

**Comment by pedro-vicente** [comment]
@czender 

on my local Win32

```
C:\nco>ncap2 -O -v -s 'erf_one=float(gsl_sf_erf(1.0));print(erf_one,"%g")' in.nc foo.nc
Command-line script line 1, column 10: unexpected character '''
```

I'll debug this later...
is there any other ncap2 run that we can try that 1) calls GSL 2) does not have any quotes "" in it?
because we don't know if this is a Win64 specific thing related to size_t or just the way Windows handles the quotes

**Comment by pedro-vicente** [comment]
@ocefpaf 

> Any reason why you chose Win-32 instead of Win-64?

just because I started building Windows code a long time ago when Win-64 did not exist and also because Win-64 binaries only work on a 64-bit machine (and Win-32 binaries work in both 32-bit and 64 bit machines)

**Comment by czender** [comment]
You can drop the ncap2 print statement completely. Other GSL examples in manual. 

I suggest adding and invoking an optional switch to CMake to turn-off ncap2 build even when all libraries are present. There is no evidence of problems with other binaries. Then we can release Conda Windows NCO with everything but ncap2 initially, and give some breathing room to solve the ncap2 issues. In fact a simpler approach may be to just turn-off GSL in CMake and release ncap2 as is. There is no evidence of problems in other ncap2 functionality. Thoughts?

**Comment by pedro-vicente** [comment]
taking the print statement out

`ncap2 -O -v -s 'erf_one=float(gsl_sf_erf(1.0))' in.nc foo.nc`

segfaults 

but changing the quotes to


```
ncap2 -O -v -s "erf_one=float(gsl_sf_erf(1.0))" in.nc foo.nc
        1 file(s) moved.

```

and then

```
ncks -v erf_one foo.nc
netcdf foo {

  variables:
    float erf_one ;

  data:
    erf_one = 0.8427008 ;

} // group /
```

so it seems the segfault is just because of the way Windows handles the quotes ?

**Comment by czender** [comment]
Progress! See https://stackoverflow.com/questions/24173825/what-does-single-quote-do-in-windows-batch-files or similar. Perhaps trial and error can be used to re-formulate the quoting in the test command so that it works in Win64. The simplest option seems to be to eliminate the print statements and change the remaining quotes from single to double:
`ncap2 -O -v -s "erf_one=float(gsl_sf_erf(1.0))" in.nc foo.nc`

**Comment by pedro-vicente** [comment]
@czender 

it seems it's mission impossible to get Windows/or the NCO ANTLR parser to accept single quotes, e.g

`ncaps -D 2 -O -v -s "erf_one=float(gsl_sf_erf(1.0));print(erf_one,'%g')" in.nc foo.nc`

from that article linked above

> However, there are Windows utilities (external commands) like WMIC that use single quotes for their own quoting purposes. But there is no standard - each utility is free to have its own interpretation of single quotes.

so, it may be possible to make the NCO ANTLR parser to accept the quote? This is a bit out of my league, it's more of a question for Henry.

other option is to make a special replacement for the single quote here only for the Windows case 

e.g instead of

`print(erf_one,'%g')`

replace the single quote with some special character , like ~

print(erf_one,~%g~)

File: recipe/run_test.bat [hunk 1]
```diff
@@ -4,5 +4,5 @@ ncgen -o in.nc in.cdl || exit 1
 
 ncks -H --trd -v one in.nc || exit 1
 
-ncap2 -O -v -s 'erf_one=float(gsl_sf_erf(1.0f));print(erf_one,"%g")' in.nc foo.nc || exit 1
-
+ncap2 -O -v -s "erf_one=float(gsl_sf_erf(1.0f));" in.nc foo.nc || exit 1
+ncks -v erf_one foo.nc || exit 1
```
Comment (ocefpaf) [inline recipe/run_test.bat]: @pedro-vicente and @czender replace the single quotes with double and removing the print, tried many escape option and quote combinations but I could not make it work, we get the expected results. See the `erf_one` value from the following `ncks` call.

TL;DR `ncap2` works as expected but users most be warned that examples from the manual may need some tweaking to make them work in the Windows CLI.

**Comment by czender** [comment]
I merged the shortened test command. We will announce the Windows quoting issue with the 4.7.1 release, and possibly develop a workaround feature in the NCO 4.7.2-4.7.3 timeframe.
